### PR TITLE
Virtual Gears

### PR DESCRIPTION
**Pull request by ericchristoffersen**
Issue #3749 

This change depends on change #3725.

### CODE REVIEW

**Comment by ericchristoffersen** [comment]
Rebased to 3725 latest.

**Comment by mattipee** [comment]
Initial reaction is that TrainSidebar.cpp is getting populated with a lot of physics and tricksy maths. Is there sense in a refactor, or tend not to in order to keep risk/change down?

File: src/Train/TrainSidebar.cpp [hunk 1]
```diff
@@ -2558,55 +2638,122 @@ void TrainSidebar::RewindLap()
...
+double IntensifyNewtons(double N, double intensity) {
+    return N * intensity;
+}
 
-        if (load <0) load = 0;
-        if (slope <-40) slope = -40;
+// When there is a divergence between simulated speed and device speed
+// adjust mass sent to device so it can emulate the simulated kinetic
+// energy.
+double TrainSidebar::getAdjustedInertialMass() const {
```
Comment (mattipee) [inline src/Train/TrainSidebar.cpp:2735]: Do we need to account also for intensity factor? I haven't made the change and tried it, but discuss....?

Comment (ericchristoffersen) [reply src/Train/TrainSidebar.cpp:2735]: The connection with IntensityFactor is that when used in anger it, by design, causes a pretty massive divergence between sim and device speed. Is why I didn't notice the inertia problem until I started using the virtual gears. 1.1 or 0.9 is small potatoes and not so noticeable but with 5x difference in modelled speed the feel problem becomes obvious.

**Comment by ericchristoffersen** [comment]
> Initial reaction is that TrainSidebar.cpp is getting populated with a lot of physics and tricksy maths. Is there sense in a refactor, or tend not to in order to keep risk/change down?

Definitely not as part of this change. I strongly prefer that a big reorganization is a functional nop.

I've been pushing for a long time to isolate training state from the trainsidebar. Is sort of absurd that a graphics element owns training state. I've been slowly isolating components from runtime state so that everything lives in trainsidebar, then eventually that state will go into its own object. Finally there will be a top level negotiator that trades data between devices, the training state and any 'future entities'.

There is a real need because some people want to add more complex controls and question is where and how that ui can work against current structure.

This is just my opinion. Plan needs to be discussed and agreed.

File: src/Train/TrainSidebar.cpp [hunk 2]
```diff
@@ -1652,16 +1656,20 @@ void TrainSidebar::guiUpdate()           // refreshes the telemetry
...
+                                                           // to within defined limits
+                    }
                 }
 
-                if (Devices[dev].type == DEV_FORTIUS || Devices[dev].type == DEV_IMAGIC) {
-	                rtData.setLoad(local.getLoad()); // and get load in case it was adjusted
-                    rtData.setSlope(local.getSlope()); // and get slope in case it was adjusted
-					// to within defined limits
-				}
+                if (!useSimulatedSpeed) {
```
Comment (ericchristoffersen) [inline src/Train/TrainSidebar.cpp:1666]: Prior to this change, fortius device speed always overwrote sim speed, even if sim speed is enabled.

**Pull request closed without merging**